### PR TITLE
Issue #2: Add New Endpoint for Historical Prices

### PR DESCRIPTION
**Pull request by asmi38**
**Goal:**
Create a new endpoint to FUSE that will allow users to query for historical prices.

**Inputs:**
Symbols (required): the stock ticker symbols that historical price data will be queried for
Date (not required, default = today): the date at which to get the historical prices for
Ranges (not required, default = 1 month): the range from the date to get the historical prices for

Note: if date and range are both present, date will supersede the range

**Function:**
IexRestController will serve as the endpoint for users of FUSE.

The method will call iexService which will have a new method getHistoricalPrices, this method will utilize the iexClient to fetch the data from the api.

The IexService getHistoricalPrices method will contain logic to call the correct method from the IexCloudClient depending on the amount of arguments passed in.

The IexCloudClient will be a new interface which will connect and make calls  to the Iex Cloud api. It will contain multiple methods which will be called by IexService.

The return value will be a list of IexHistoricalPrice objects which will contain the specified data for each day.

**Tests:**
The unit tests that will be conducted will check to see if the historicalData lines up with known dates and values and will work for varying inputs:

- Symbol, no date, no range
- Symbol, date, no range
- Symbol, no date, range
- No symbol input

### CODE REVIEW

File: src/main/java/org/galatea/starter/service/IexClient.java [hunk 1]
```diff
@@ -1,6 +1,7 @@
 package org.galatea.starter.service;
 
 import java.util.List;
+import org.galatea.starter.domain.IexHistoricalPrice;
```
Comment (grundygolem2) [inline src/main/java/org/galatea/starter/service/IexClient.java]: Don't import classes or methods you don't use

File: src/test/java/org/galatea/starter/entrypoint/IexRestControllerTest.java [hunk 2]
```diff
@@ -81,4 +81,88 @@ public void testGetLastTradedPriceEmpty() throws Exception {
         .andExpect(jsonPath("$", is(Collections.emptyList())))
         .andReturn();
   }
+
+  @Test
+  public void testGetHistoricalPriceSymbol() throws Exception {
+
+    MvcResult result = this.mvc.perform(
+        org.springframework.test.web.servlet.request.MockMvcRequestBuilders
```
Comment (grundygolem2) [inline src/test/java/org/galatea/starter/entrypoint/IexRestControllerTest.java]: Can you avoid having the full path here?

File: src/main/java/org/galatea/starter/service/IexService.java [hunk 3]
```diff
@@ -45,5 +47,27 @@
...
+   * @param date the date to begin getting historical prices at.
+   * @param range the range to get historical prices at.
+   * @return a list of historical prices for the specified symbol and time frame.
+   */
+  public List<IexHistoricalPrice> getHistoricalPrices(
+      final String symbol,
+      final String range,
+      final String date) {
+    if (symbol.isEmpty()) {
+      return Collections.emptyList();
```
Comment (grundygolem2) [inline src/main/java/org/galatea/starter/service/IexService.java]: Given that we expect a symbol in our endpoint, is emptyList the correct response here?
(not being snarky, genuine question)

Comment (asmi38) [reply src/main/java/org/galatea/starter/service/IexService.java]: This was to catch the scenario where a request is made that has the symbol parameter, but the value is empty, for example, this testcase:
```
        org.springframework.test.web.servlet.request.MockMvcRequestBuilders
            .get("/iex/historicalPrice?symbol=") 
```
It would satisfy the required condition of the @RequestParam symbol parameter in the IexRestController, but would be unable to make a successful request to the IEX API as the symbol is a required parameter.

A few alternative solutions would be to make the if elses more explicit and to return an emptyList as a base return case. 

I could also make the symbol @RequestParam not required and test for null but I don't think that would be an effective work around.

Another option would be to do the check in the IexRestController but I think it would be best to keep the logic in the Service class.

With this in mind, do you think I should still change it? Or perhaps there's a better alternative I'm missing

Comment (grundygolem2) [reply src/main/java/org/galatea/starter/service/IexService.java]: Gotcha, that makes sense. In that case, you could look into adding request validation, because symbol= isn't really a valid request from the client. There are annotation based methods you could use to verify that symbol is populated.

**Review by grundygolem2 — APPROVED** [review]
Looks Good

File: src/test/java/org/galatea/starter/entrypoint/IexRestControllerTest.java [hunk 4]
```diff
@@ -74,11 +75,94 @@ public void testGetLastTradedPrice() throws Exception {
...
+  @Test
+  public void testGetHistoricalPriceEmpty() throws Exception {
+
+    MvcResult result = this.mvc.perform(
+        MockMvcRequestBuilders
+            .get("/iex/historicalPrice?symbol=")
+            .accept(MediaType.APPLICATION_JSON_VALUE))
+        .andExpect(status().isBadRequest())
+        .andReturn();
+  }
```
Comment (alex-power) [inline src/test/java/org/galatea/starter/entrypoint/IexRestControllerTest.java:167]: Please add one more test to validate the case if the IEX server is down or fails.

**Review by alex-power — CHANGES_REQUESTED** [review]
Nice work. Please add one more test case per my comment.

File: src/main/java/org/galatea/starter/entrypoint/RestExceptionHandler.java [hunk 5]
```diff
@@ -73,8 +74,18 @@
     return buildResponseEntity(error);
   }
 
+  @ExceptionHandler(FeignException.class)
+  protected ResponseEntity<Object> handleFeignStatusException(final FeignException exception) {
+    log.debug("Error connecting to IEX", exception);
```
Comment (alex-power) [inline src/main/java/org/galatea/starter/entrypoint/RestExceptionHandler.java]: Errors are not debug statements, should be ```log.error```

**Review by alex-power — CHANGES_REQUESTED** [review]

**Review by alex-power — APPROVED** [review]
